examples/chat/src/models/message.rs

```diff
@@ -0,0 +1,113 @@
+use chrono::NaiveDateTime;
+use diesel::deserialize::FromSqlRow;
+use diesel::dsl::{And, AsSelect, Desc, Eq, Lt};
+use diesel::helper_types::{InnerJoin, Select};
+use diesel::pg::Pg;
+use diesel::prelude::*;
+use diesel::row::Row;
+use diesel::sql_types::Bool;
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+use super::thread::Thread;
+use super::user::User;
+use crate::schema::messages::{self, dsl as col};
+use crate::schema::{threads, users};
+
+pub type TableWithJoins = InnerJoin<InnerJoin<messages::table, threads::table>, users::table>;
+pub type DefaultSelection = (
+    AsSelect<Message, Pg>,
+    AsSelect<Thread, Pg>,
+    AsSelect<User, Pg>,
+);
+
+#[derive(Associations, Queryable, Selectable, Serialize)]
+#[diesel(belongs_to(Thread))]
+#[diesel(belongs_to(User, foreign_key = author_id))]
+#[diesel(table_name = messages)]
+#[diesel(check_for_backend(Pg))]
+#[serde(rename_all = "camelCase")]
+pub struct Message {
+    id: Uuid,
+
+    body: String,
+
+    #[serde(skip)]
+    author_id: Uuid,
+
+    #[serde(skip)]
+    thread_id: Uuid,
+
+    created_at: NaiveDateTime,
+    updated_at: NaiveDateTime,
+}
+
+#[derive(AsChangeset, Deserialize)]
+#[diesel(table_name = messages)]
+pub struct MessageChangeSet {
+    pub body: String,
+}
+
+#[derive(Deserialize, Insertable)]
+#[diesel(table_name = messages)]
+#[serde(rename_all = "camelCase")]
+pub struct MessageParams {
+    pub body: String,
+
+    pub author_id: Option<Uuid>,
+    pub thread_id: Option<Uuid>,
+}
+
+#[derive(Serialize)]
+pub struct MessageWithJoins {
+    #[serde(flatten)]
+    message: Message,
+    thread: Thread,
+    author: User,
+}
+
+pub fn by_author(id: Uuid) -> Eq<col::author_id, Uuid> {
+    col::author_id.eq(id)
+}
+
+pub fn by_cursor(
+    cursor: (NaiveDateTime, Uuid),
+) -> And<Eq<col::created_at, NaiveDateTime>, Lt<col::id, Uuid>, Bool> {
```

Comment:
These return types are a bit gnarly but it beats loading the schema DSL for the table in the routes module and turning your editor into a big ol' bag of skittles.

They look like named rails scopes on a model but they are composable and do not execute queries. It's a pretty cool pattern and makes me happy that I decided to use an ORM.